### DISCUSSION TITLE
How to send message from glsp-server to glsp-theia

### DISCUSSION THREAD

**Discussion by wss29** [eclipse-glsp/glsp#1175] [Q&A] [answered]
hi,
   I have a service in my glsp-server,  I want to send some messages from glsp-server to glsp-thiea in this service, and display this message in the message box in theia. I am thinking of using actionDispatcher, but how could I handle dispatcher action in glsp-theia? Could anyone give me some ideas?

**Comment by tortmayr** [comment, marked as answer]
Hi @wss29,

the GLSP protocol allows you to send `MessageActions` and the Theia integration has a default handling for this messages in place that automatically forwards to Theia`s message service and displays them in the message box.
The message action definition looks like this: 
```ts
export interface MessageAction extends Action {
    kind: typeof MessageAction.KIND;

    severity: SeverityLevel;

    /**
     * The message that shall be shown to the user.
     */
    message: string;

    /**
     * Further details on the message.
     */
    details?: string;
}
```
The severity level determines whether the message is printed as info, warning or error, if details are passed an additional "Show Details" button will be displayed which opens a dialog with the details content when clicked.

So you can simply dispatch a new `MessageAction` on the server and the  glsp-theia-integration will automatically forward it to Theia's message service.

If this default handling is not enough for you, you could implement a custom action and handler on the client side similar to https://github.com/eclipse-glsp/glsp-theia-integration/blob/refs/tags/v2.0.0/packages/theia-integration/src/browser/diagram/features/notification/theia-glsp-message-service.ts
